experimental/parser/parse_decl.go

```diff
@@ -0,0 +1,695 @@
+				break
+			}
+		} else if args.Equals.Nil() {
+			args.Equals = eq
+		}
+
+		expr := parseExpr(p, c, taxa.Def.In())
+		if expr.Nil() {
+			continue // parseExpr already generated diagnostics.
+		}
+
+		exprIs := taxa.FieldTag
+		if kw.Text() == "option" {
+			exprIs = taxa.OptionValue
+		} else if ty.Nil() {
+			exprIs = taxa.EnumValue
+		}
+
+		if !args.Value.Nil() {
+			p.Error(errMoreThanOne{
+				first: args.Value,
+				// This join will make it so that the = is included in the
+				// "consider removing this" suggestion.
+				second: report.Join(eq, expr),
+				what:   exprIs,
+			})
+			continue
+		}
+
+		args.Value = expr
+		switch {
+		case args.Equals.Nil():
+			p.Error(errUnexpected{
+				what:  expr,
+				where: taxa.Equals.Without(),
+				prev:  args.Equals,
+				got:   exprIs,
+			})
+		case !args.Options.Nil():
+			p.Error(errUnexpected{
+				what:  expr,
+				where: taxa.CompactOptions.After(),
+				prev:  args.Options,
+				got:   exprIs,
+			})
+		case !braces.Nil():
+			p.Error(errUnexpected{
+				what:  expr,
+				where: taxa.Body.After(),
+				prev:  braces,
+				got:   exprIs,
+			})
+		}
+	}
+
+	// If we didn't see any braces, this def needs to be ended by a semicolon.
+	if braces.Nil() {
+		semi, err := p.Punct(c, ";", taxa.Def.After())
+		args.Semicolon = semi
+		if err != nil {
+			p.Error(err)
+		}
+	}
+
+	// Convert something that looks like an enum value into one.
+	if tyPath := ty.AsPath(); !tyPath.Nil() && path.Nil() &&
+		// The reason for this is because if the user writes `message {}`, we want
+		// to *not* turn it into an enum value with a body.
+		//
+		// TODO: Add a case for making sure that `rpc(foo)` is properly
+		// diagnosed as an anonymous method.
+		braces.Nil() {
+		args.Name = tyPath.Path
+		args.Type = ast.TypeAny{}
+	}
+
+	def := p.NewDeclDef(args)
+
+	if !inputs.Nil() {
+		parseTypeList(p, inputs, def.WithSignature().Inputs(), taxa.MethodIns)
+	}
+	if !outputs.Nil() {
+		parseTypeList(p, outputs, def.WithSignature().Outputs(), taxa.MethodOuts)
+	} else if !outputTy.Nil() {
+		p.Errorf("missing `(...)` around method return type").With(
+			report.Snippetf(outputTy, "help: replace this with `(%s)`", outputTy.Span().Text()),
+		)
+		def.WithSignature().Outputs().Append(outputTy)
+	}
+
+	if !braces.Nil() {
+		var in taxa.Subject
+		switch kw.Text() {
+		case "message":
+			in = taxa.Message
+		case "enum":
+			in = taxa.Enum
+		case "service":
+			in = taxa.Service
+		case "extend":
+			in = taxa.Extend
+		case "group":
+			in = taxa.Field
+		case "oneof":
+			in = taxa.Oneof
+		case "rpc":
+			in = taxa.Method
+		default:
+			in = taxa.Def
+		}
+
+		def.SetBody(parseBody(p, braces, in))
+	}
+
+	return def
+}
+
+// parseTypeList parses a type list out of a bracket token.
+func parseTypeList(p *parser, parens token.Token, types ast.TypeList, in taxa.Subject) {
+	tys := commas(p, parens.Children(), true, func(c *token.Cursor) (ast.TypeAny, bool) {
+		ty, _ := parseType(p, c, in.In(), false)
```

Comment:
It's very hard to separate these two functions. What I could do is provide two functions that call the common implementation with this bool set; the alternative is duplicating the fairly fussy type parser logic in two functions.

Reply:
Maybe name this one `parseTypeAndMaybePath` and then also have `parseType` (used here) and `parseTypeAndPath` (used in `parseDecl`) that delegate to it. I think that would make the various call sites (like this) easier to understand -- having method calls with boolean params like this can be awkward unless the method name makes it super obvious what the bool value means.